prover/lang/kore-lang.md

```diff
@@ -260,6 +260,34 @@ module KORE-HELPERS
 ```
 
 ```k
+  syntax Bool ::= isFunctional(GoalId, Symbol) [function]
+
+  rule isFunctional(_, plus #Or minus #Or select #Or union
+         #Or singleton #Or emptyset) => true
+
+  rule [[ isFunctional(_, S) => true ]]
+       <declaration> axiom _: functional(S) </declaration>
+  rule [[ isFunctional(GId, S) => true ]]
+       <id> GId </id>
+       <local-decl> axiom _: functional(S) </local-decl>
+  rule isFunctional(_, S) => false [owise]
+
+  syntax Bool ::= isSurelyFunctional(GoalId, Pattern) [function]
+  rule isSurelyFunctional(GId, P)
+    => areSurelyFunctional(GId, P, .Patterns)
+
+  syntax Bool ::= areSurelyFunctional(GoalId, Patterns) [function]
+
+  rule areSurelyFunctional(_, .Patterns) => true
+  rule areSurelyFunctional(GId, I:Int, Ps)
+    => areSurelyFunctional(GId, Ps)
+  rule areSurelyFunctional(GId, V{S}, Ps)
+    => areSurelyFunctional(GId, Ps)
+  rule areSurelyFunctional(GId, R ( ARGS ), Ps)
+    => isFunctional(GId, R) andBool
+       areSurelyFunctional(GId, ARGS ++Patterns Ps)
+  rule areSurelyFunctional(_, _) => false [owise]
```

Comment:
Whats the reason that `isFunctional` and `isSurelyFunctional` are separate? Since they're both do best-effort syntactical checking I'm not sure we need both. Lets merge them into `isFunctional`?